datasimulator/graph.py

```diff
@@ -0,0 +1,265 @@
+from generator import generate_list_numbers
+
+EXCLUDED_NODE = ["program", "root", "data_release"]
+
+
+class Graph(object):
+    """
+    Graph representation class
+    """
+
+    def __init__(self, dictionary, program, project):
+        """
+        Graph constructor
+
+        Args:
+            dictionary(DataDictionary): a dictionary instance
+            program(str): program name
+            project(str): project name
+
+        Outputs:
+            None
+        """
+        self.dictionary = dictionary
+        self.root = None
+        self.program = program
+        self.project = project
+        self.nodes = []
+
+    def prelimary_dictionary_check(self):
+        """
+        raise exception if dictionary has not initialized yet
+        """
+        if self.dictionary is None:
+            raise UserError("Dictionary is not initialized!!!")
+        return True
+
+    def _get_list_of_node_names(self):
+        """
+        return a list of node names
+        """
+        return [k for k in self.dictionary.schema if k not in EXCLUDED_NODE]
+
+    def generate_nodes_from_dictionary(self):
+        """
+        generate nodes from dictionary
+
+        """
+        # logger.info('Start simulating data')
+        for node_name in self._get_list_of_node_names():
+            node = Node(node_name, self.dictionary.schema[node_name], self.project)
+            if node_name == "project":
+                self.root = node
+            self.nodes.append(node)
+
+    def get_node_with_name(self, node_name):
+        """
+        get node object given name
+
+        Args:
+            node_name(str): node name
+
+        Outputs:
+            Node: node object
+        """
+        for node in self.nodes:
+            if node.name == node_name:
+                return node
+        return None
+
+    def _add_required_link_to_node(
+        self, node, link_node_name, link_name, multiplicity=None, skip=True
+    ):
+        """
+        assign required links to a node
+
+        Args:
+            node(Node): node object
+            link_node_name(str): link node name
+            multiplicity(str): link type (one_to_one, one_to_many, ..etc.)
+            skip(bool): skip raising an exception to terminate
+
+        Outputs:
+            None or raise exception
+
+        """
+        # skip all the links to Project node
+        if link_node_name in EXCLUDED_NODE:
+            return
+        node_parent = self.get_node_with_name(link_node_name)
+
+        if not node_parent:
+            msg = "Node {} have a link to node {} which does not exist".format(
+                    node.name, link_node_name
+                )
+            if skip:
+                logger.error(msg)
+            else:
+                raise DictionaryError(message=msg)
+
+        node.required_links.append(
+            {"node": node_parent, "multiplicity": multiplicity, "name": link_name}
+        )
+
+    def graph_validation(self, required_only=False):
```

Comment:
There's a fair amount of error checking on the node links that happens in `construct_graph_edges` (and elsewhere), could maybe expand this function to include all of the error handling and assume validity in those other functions?